credentials/credentials.go

```diff
@@ -98,19 +94,42 @@ func (c *tlsCreds) GetRequestMetadata(ctx context.Context) (map[string]string, e
 	return nil, nil
 }
 
-func (c *tlsCreds) DialWithDialer(dialer *net.Dialer, network, addr string) (_ net.Conn, err error) {
+type timeoutError struct{}
+
+func (timeoutError) Error() string   { return "credentials: Dial timed out" }
+func (timeoutError) Timeout() bool   { return true }
+func (timeoutError) Temporary() bool { return true }
+
+func (c *tlsCreds) Handshake(addr string, rawConn net.Conn, timeout time.Duration) (_ net.Conn, err error) {
+	// borrow some code from tls.DialWithDialer
+	var errChannel chan error
+	if timeout != 0 {
+		errChannel = make(chan error, 2)
+		time.AfterFunc(timeout, func() {
+			errChannel <- timeoutError{}
+		})
+	}
 	if c.config.ServerName == "" {
-		c.config.ServerName, _, err = net.SplitHostPort(addr)
-		if err != nil {
-			return nil, fmt.Errorf("credentials: failed to parse server address %v", err)
+		colonPos := strings.LastIndex(addr, ":")
+		if colonPos == -1 {
+			colonPos = len(addr)
 		}
+		c.config.ServerName = addr[:colonPos]
 	}
-	return tls.DialWithDialer(dialer, network, addr, &c.config)
-}
-
-// Dial connects to addr and performs TLS handshake.
-func (c *tlsCreds) Dial(network, addr string) (_ net.Conn, err error) {
-	return c.DialWithDialer(new(net.Dialer), network, addr)
+	conn := tls.Client(rawConn, &c.config)
+	if timeout == 0 {
```

Comment:
It's fine to have a goroutine even if it's not strictly necessary here. It'll make the code simpler to write

```go
go func() {
  errChannel <- conn.Handshake()
}()
if err := <-errChannel; err != nil {
  rawConn.Close()
  return nil, err
}
return conn, nil
```